File: app/src/main/java/com/juniperphoton/myersplash/utils/NotificationUtil.java
package com.juniperphoton.myersplash.utils;

import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.WallpaperManager;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.support.v4.app.NotificationCompat;
import android.support.v4.app.TaskStackBuilder;
import android.support.v4.content.FileProvider;
import android.util.SparseArray;

import com.juniperphoton.myersplash.R;
import com.juniperphoton.myersplash.base.App;
import com.juniperphoton.myersplash.service.BackgroundDownloadService;

import java.io.File;
import java.util.HashMap;

public class NotificationUtil {

    private static int mLastId = 0;
    public static int NOT_ALLOCATED_ID = -10000;
    private static HashMap<Uri, Integer> uriHashMap = new HashMap<>();
    private static SparseArray<NotificationCompat.Builder> integerBuilderHashMap = new SparseArray<>();

    private static int findNIdByUri(Uri downloadUri) {
        int nId = NOT_ALLOCATED_ID;
        if (uriHashMap.containsKey(downloadUri)) {
            nId = uriHashMap.get(downloadUri);
        }
        return nId;
    }

    private static NotificationCompat.Builder findBuilderById(int id) {
        if (integerBuilderHashMap.get(id) != null) {
            return integerBuilderHashMap.get(id);
        }
        return null;
    }

    private static NotificationManager getNotificationManager() {
        return (NotificationManager) App.getInstance().getSystemService(Context.NOTIFICATION_SERVICE);
    }

    public static void cancelNotification(Uri downloadUri) {
        int nId = findNIdByUri(downloadUri);
        if (nId != NOT_ALLOCATED_ID) {
            getNotificationManager().cancel(nId);
        }
    }

    public static void cancelNotificationById(int nId) {
        if (nId != NOT_ALLOCATED_ID) {
            getNotificationManager().cancel(nId);
        }
    }

    public static void showErrorNotification(Uri downloadUri, String fileName, String url) {
        int nId;
        nId = findNIdByUri(downloadUri);
        if (nId == NOT_ALLOCATED_ID) {
            uriHashMap.put(downloadUri, mLastId);
            nId = mLastId;
            mLastId++;
        }

        Intent intent = new Intent(App.getInstance(), BackgroundDownloadService.class);
        intent.putExtra("NAME", fileName);
        intent.putExtra("URI", url);
        intent.putExtra("NID", nId);

        PendingIntent resultPendingIntent = PendingIntent.getService(App.getInstance(), 0, intent, 0);

        NotificationCompat.Builder builder = new NotificationCompat.Builder(App.getInstance())
                .setContentTitle("Download error.")
                .setContentText("Please checkAndRequest your network and retry.")
                .setSmallIcon(R.drawable.ic_cancel_white_36dp);

        builder.addAction(R.drawable.ic_replay_white_48dp, "Retry", resultPendingIntent);

        getNotificationManager().notify(nId, builder.build());
    }

    public static void showCompleteNotification(Uri downloadUri, Uri fileUri) {
        NotificationCompat.Builder mBuilder = new NotificationCompat.Builder(App.getInstance())
                .setContentTitle("Saved :D")
                .setContentText("Tap to crop and set as wallpaper.")
                .setSmallIcon(R.drawable.small_icon);

        File file = new File(fileUri.getPath());
        Uri uri = FileProvider.getUriForFile(App.getInstance(), App.getInstance().getString(R.string.authorities), file);
        Intent intent = WallpaperManager.getInstance(App.getInstance()).getCropAndSetWallpaperIntent(uri);

        TaskStackBuilder stackBuilder = TaskStackBuilder.create(App.getInstance());
        stackBuilder.addNextIntent(intent);

        PendingIntent resultPendingIntent =
                stackBuilder.getPendingIntent(0, PendingIntent.FLAG_UPDATE_CURRENT);
        mBuilder.setContentIntent(resultPendingIntent);

        int nId;
        nId = findNIdByUri(downloadUri);
        if (nId != NOT_ALLOCATED_ID) {
            getNotificationManager().notify(nId, mBuilder.build());
        }
    }

    public static void showProgressNotification(String title, String content, int progress, Uri downloadUri) {
        int nId;
        nId = findNIdByUri(downloadUri);
        if (nId == NOT_ALLOCATED_ID) {
            uriHashMap.put(downloadUri, mLastId);
            nId = mLastId;
            mLastId++;
        }

        NotificationCompat.Builder builder = findBuilderById(nId);
        if (builder == null) {
            builder = new NotificationCompat.Builder(App.getInstance())
                    .setContentTitle(title)
                    .setContentText(content)
                    .setSmallIcon(R.drawable.download_small_icon);
            integerBuilderHashMap.put(nId, builder);
        } else {
            builder.setProgress(100, progress, false);
        }
        getNotificationManager().notify(nId, builder.build());
    }
}

File: app/src/main/java/com/juniperphoton/myersplash/utils/NotificationUtil.kt
package com.juniperphoton.myersplash.utils

import android.app.NotificationManager
import android.app.PendingIntent
import android.app.PendingIntent.FLAG_UPDATE_CURRENT
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.support.v4.app.NotificationCompat
import android.support.v4.app.TaskStackBuilder
import android.util.SparseArray
import com.juniperphoton.myersplash.App
import com.juniperphoton.myersplash.R
import com.juniperphoton.myersplash.activity.ManageDownloadActivity
import com.juniperphoton.myersplash.service.BackgroundDownloadService
import java.util.*

class NotificationUtil {
    companion object {
        private var mLastId = 0
        var NOT_ALLOCATED_ID = -10000
        private val uriHashMap = HashMap<Uri, Int>()
        private val integerBuilderHashMap = SparseArray<NotificationCompat.Builder>()

        private fun findNIdByUri(downloadUri: Uri): Int {
            var nId: Int = NOT_ALLOCATED_ID
            if (uriHashMap.containsKey(downloadUri)) {
                nId = uriHashMap[downloadUri] ?: -1
            }
            return nId
        }

        private fun findBuilderById(id: Int): NotificationCompat.Builder? {
            if (integerBuilderHashMap.get(id) != null) {
                return integerBuilderHashMap.get(id)
            }
            return null
        }

        private val notificationManager: NotificationManager
            get() = App.instance.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

        fun cancelNotification(downloadUri: Uri) {
            val nId = findNIdByUri(downloadUri)
            if (nId != NOT_ALLOCATED_ID) {
                notificationManager.cancel(nId)
            }
        }

        fun cancelNotificationById(nId: Int) {
            if (nId != NOT_ALLOCATED_ID) {
                notificationManager.cancel(nId)
            }
        }

        fun showErrorNotification(downloadUri: Uri, fileName: String, url: String) {
            var nId: Int
            nId = findNIdByUri(downloadUri)
            if (nId == NOT_ALLOCATED_ID) {
                uriHashMap.put(downloadUri, mLastId)
                nId = mLastId
                mLastId++
            }

            val intent = Intent(App.instance, BackgroundDownloadService::class.java)
            intent.putExtra(Params.NAME_KEY, fileName)
            intent.putExtra(Params.URL_KEY, url)
            intent.putExtra(Params.CANCEL_NID_KEY, nId)

            val resultPendingIntent = PendingIntent.getService(App.instance, 0, intent, 0)

            val builder = NotificationCompat.Builder(App.instance)
                    .setContentTitle(App.instance.getString(R.string.download_error))
                    .setContentText(App.instance.getString(R.string.download_error_retry))
                    .setSmallIcon(R.drawable.vector_ic_clear_white)

            builder.addAction(R.drawable.ic_replay_white_48dp, App.instance.getString(R.string.retry_act),
                    resultPendingIntent)

            notificationManager.notify(nId, builder.build())
        }

        fun showCompleteNotification(downloadUri: Uri, fileUri: Uri) {
            val nId: Int = findNIdByUri(downloadUri)

            val builder = NotificationCompat.Builder(App.instance)
                    .setContentTitle(App.instance.getString(R.string.saved))
                    .setContentText(App.instance.getString(R.string.tap_to_open_manage))
                    .setSmallIcon(R.drawable.small_download_ok)

            //File file = new File(fileUri.getPath());
            //Uri uri = FileProvider.getUriForFile(App.instance, App.instance.getString(R.string.authorities), file);
            //Intent intent =  WallpaperManager.getInstance(App.instance).getCropAndSetWallpaperIntent(uri);
            injectIntent(builder)

            if (nId != NOT_ALLOCATED_ID) {
                notificationManager.notify(nId, builder.build())
            }
        }

        fun showProgressNotification(title: String, content: String, progress: Int,
                                     filePath: String, downloadUri: Uri) {
            var nId: Int
            nId = findNIdByUri(downloadUri)
            if (nId == NOT_ALLOCATED_ID) {
                uriHashMap.put(downloadUri, mLastId)
                nId = mLastId
                mLastId++
            }

            var builder = findBuilderById(nId)
            if (builder == null) {
                builder = NotificationCompat.Builder(App.instance)
                        .setContentTitle(title)
                        .setContentText(content)
                        .setSmallIcon(R.drawable.vector_ic_file_download)
                integerBuilderHashMap.put(nId, builder)
            } else {
                builder.setProgress(100, progress, false)
            }
            injectIntent(builder)
            notificationManager.notify(nId, builder!!.build())
        }

        private fun injectIntent(builder: NotificationCompat.Builder?) {
            val intent = Intent(App.instance, ManageDownloadActivity::class.java)
            val stackBuilder = TaskStackBuilder.create(App.instance)
            stackBuilder.addNextIntent(intent)

            val resultPendingIntent = stackBuilder.getPendingIntent(0, FLAG_UPDATE_CURRENT)
            builder?.setContentIntent(resultPendingIntent)
        }
    }

}
